hawkmoth/__init__.py

```diff
@@ -151,12 +130,48 @@ def _get_filenames(self):
         """
         raise NotImplementedError(self.__class__.__name__ + '._get_filenames')
 
+    def __get_all_filenames(self):
+        nfiles = 0
+
+        filenames = glob.glob(os.path.join(self.env.config.cautodoc_root, '*.[ch]'))
+        filenames += glob.glob(os.path.join(self.env.config.cautodoc_root, '**/*.[ch]'))
```

Comment:
If I understand this correctly, encountering a directive without filenames specified will scan all the `*.[ch]` files in source tree. For every directive.

The obvious improvement would be to scan all the files only once at extension setup. However, I still have really mixed feelings about on the one hand restricting to `*.[ch]` suffix and on the other hand scanning the entire source tree, and parsing all the files if there's just one directive without a filename specified.

Say, if you have a moderately big project, but only have a handful of files to document, you'd end up running clang on every file in the project across the entire project.

Reply:
> If I understand this correctly, encountering a directive without filenames specified will scan all the `*.[ch]` files in source tree. For every directive.

No, we take the file names on every directive, which might be a bit wasteful, but it's just walking the file system here. Clang will only be called once per file (not considering the issue you raised with parallel builds).

That said, we can take the file list only once though for a small performance improvement indeed.

> The obvious improvement would be to scan all the files only once at extension setup.

At setup we don't know if there are any directives that do not list a file name. Essentially, this means that we will always have worst case parsing even if the user never uses the auto discovery explicitly, which would be against your last point and which I'm also against.

This may be worked around if we instead have to pass all the filenames we're interested in through the configuration file instead of a directory. Or maybe a white / black list of source files.

And as for the globing pattern(s), we could default to `*.[ch]` but make it user configurable.

> However, I still have really mixed feelings about on the one hand restricting to `*.[ch]` suffix and on the other hand scanning the entire source tree, and parsing all the files if there's just one directive without a filename specified.
> 
> Say, if you have a moderately big project, but only have a handful of files to document, you'd end up running clang on every file in the project across the entire project.

Hmm, I don't think we can have it both ways, but that's why the documentation should be very clear that omitting the file names will likely trade performance for convenience.

Reply:
> > Say, if you have a moderately big project, but only have a handful of files to document, you'd end up running clang on every file in the project across the entire project.
> 
> Hmm, I don't think we can have it both ways, but that's why the documentation should be very clear that omitting the file names will likely trade performance for convenience.

Well, I guess a possible compromise would be to stop parsing when we get a match. But if anyone is ever going to rely on auto discovery, they should be assured that there is no ambiguity in their docs. And returning the 1st hit might miss the fact that there are multiple symbols with the same name. And the likelihood of such duplicate symbol names only increase with project size.